### DISCUSSION TITLE
Added: LDM Notes for August 24th, 2020

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#3818] [LDM Notes]
https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-08-24.md

Agenda:

- Warnings on types named `record`
- `base` calls on parameterless `record`s
- Omitting unnecessary synthesized `record` members
- [`record` `ToString` behavior review](https://github.com/dotnet/csharplang/blob/master/proposals/csharp-9.0/records.md#printing-members-printmembers-and-tostring-methods)
    - Behavior of trailing commas
    - Handling stack overflows
    - Should we omit the implementation of `ToString` on `abstract` records
    - Should we call `ToString` prior to `StringBuilder.Append` on value types
    - Should we try and avoid the double-space in an empty record
    - Should we try and make the typename header print more economic
- Reference equality short circuiting

**Comment by PathogenDavid** [comment]
> We could theoretically omit providing an implementation of this method on `abstract` record types as it will never be called by the derived `ToString` implementation.

Not that this changes the conclusion, but once classes can be derived from records it could potentially be used by a derived class.

> ### Reference equality short circuiting
> We will do this. 

🎉

**Comment by MgSam** [comment]
[202 ](https://en.wikipedia.org/wiki/202) was a good year for both C# LDMs and the Pantheon.

EDIT: Here are the meeting notes in the original Latin for anyone interested:
<details>
  <summary>Lingua C # testimonii ad Design July 27th, MMXX</summary>
Publica
Admonitiones recordum in nomine types
basis vocat in parameterless records
Omissis necesse summatim perstringitur recordum membra
tostring recordum mores review
Mores trailing de commatibus
BIBLIOTHECA tractantem redundat
Si autem omittamus implementation tostring in abstracto records
Si prior ad vocamus tostring StringBuilder.Append ad valorem types
Conantur vitare debemus vacuo spatio duplex testimonium
Magis debemus conari ut aliquet typename caput print
Reference aequalitatem breviter circuiting
Sponsors
"Qui regis pependit in suis principiis"

"Depinxit oculos meos et omnem turrim eburneam".

disputationem
Admonitiones recordum in nomine types
De quibus antea nos fecit admonitionem in C # IX est genus, ubi nominatur recordum, ut ea fieri necesse definitiones fugit in agrum ut bene parse. Hoc fractionis mutatione facta in C # IX, sed non expressis verbis recordarentur exitus ad disputationem de monuerit. Nos quoque agatur de quibus adhuc hic potuimus, ut casus inferioribus genus de nominibus est grande iuxta exemplum fluctus, sed ex hypothesi solum boni faciunt lowercase nominibus alia. Eg si esset denuntiatum satus in genere et nomine async var. Absit qui volunt in usus eorum features habent codebase solution in analyzers.

conclusioni
Nos mos moneo in types nomine recordum in C # AG9 Praeterea disputationem erit opus fiet in lowercase generis nomina generatim, ut sunt in globum compactionis considerations, et sentire quae disallowing recordum esset bonum fumus test nam receptio est idea in generali.

basis vocat in parameterless records
Habemus parvi foraminis, in qua id non iure records hodie et non est in habemus quo iure facit:

recordum Base (int a);
recordarentur Ex aequatione Base (I);
In amet enim hodie sicut repugnat expressis verbis: «Est error sit parumper record providere record_base argument_list record_declaration si non continet in parameter_list". Est autem ratio permittit, dum consilio ceciderint de principiis eius profectus postquam defaltam constructor et nominis non est agnitum prius conditor Hoc videtur contra principia circa plerumque faciebat types recordum minor. Praeterea, sicut scriptum est in amet hodie est non potest operari circa hoc dabit proventus per Related album est inanis modularis, cum modulus lists licet non sit inanis.

Hoc problema est relative simplex solutio: ut nos in default est prima recordum exprimentem quid nominis conditor et conditor, nisi provisum esset alius artifex. Hoc enim videtur bene reticulum existentium praecepta et consilia pro futuro bene operari videtur primaria communis constructors. II considerans esset Fortunam est nobis exigendo inanis parens omnium declarationem in casibus genus, vel solam in quibus casibus aliter tolli posse ut default conditor (si addita est conditor aliud recordum per definitionem).

conclusioni
Non te vana patitur quasi parens et conditor in C # 9. erant 'interested in permittens parens et conditor default neglegenda cum esset secus taliter emittatur, sed non facit nisi id in C # IX est ex tempore exigentiis deinde finiuit repellendum esse v.

Omissis necesse summatim perstringitur recordum membra
Ad quartum dicendum quod in certa aliqua recordum possit sodales missionem si erant 'Unneeded. Nam eget neque obsignatum MEMORIAE trahit System.Object EqualityContract res quasi ex contractu sola. Et simpliciorem de tostring implementation potuimus scire, si nos non erit ibi puer vocare debeo, quia types PrintMembers.

Autem, obstante ducens ad simpliciores emittitur codice enim haec types, hoc faciam erant, quibus cum alio codice facti sunt magis complicated. Nos opus habere ut ad rem magis complicated rationem in his casibus, compiler. Praeterea, inter se occurrunt, ut apud fontem generantibus buy instrumenta et alia ratione si hoc opus habes duplicare records programmatically erit recordum ut penitus per aequalitatem vel alia de eo codice "simpliciores" hac mechanism. Non enim adhuc non video verum metadata pereat luxoriosa suis de hac sem, ut nomina non de agris recordum declarationem.

conclusioni
Sunt magis verisimile est downsides upsides ad aliquid aliud faciunt hic. Pergo habere nos hic esse iusto.

tostring recordum mores review
Nunc habemus a record tostring rogationem et implementation pro fastigiis stantes ut youll ire per aliquas quaestiones ab initial implementation recensionem. Prior area huius artis inventa est in C # quod pertinet ad factum sive tostring tuples et anonymous types.

Mores trailing de commatibus
Est rogationem ad simpliciorem reddere elit, semper in tostring excudendi trailing commatibus post proprietatibus. Hoc opus ad aliquid removendum bool tracking num flaccescit in acutum a C Diatonico ex parte insigniori typis mandatum a parentis tostring vocationem. Sed nec prior es (vel in aliis linguis et C #) ut potuimus invenire hoc facit. Praeterea, inter quadrigis membra eius iurgia provocat dextrum reactionem Sententia, ut imperfectum et ut nobis esset fama post primam cimex shipping est.

conclusioni
Non trailing commatibus.

BIBLIOTHECA tractantem redundat
Est autem quaedam de tostring si non derivabatur ad genus record est recursive. Dum haec sit vera aequalitas matris et radicis hashCode modi conlata in libros sit aliquid spectat quod tostring esset aliquantulus speciale, quod est instrumentum illud a elit would utor ut a print-ex censu ut in cyclica elementum in primus est. Non est in JavaScript exempla huius mundi parte esse dolorem. Sed quis erit solution quod records et ascendit cum limitations, quia non possunt custodire contra mediate per circulos qui sunt records proprietatibus, non tamen tradit eiusdem generis aut in circuitus, proprietatibus, quae omnino non sunt records . In generali, si rem in acta referat a user notitia type cum cyclica structuram, haec esse a forsit non iens ut have ut iudicio contendam adversum iam hashCode aequalitatem et.

conclusioni
Nos mos nunquam tentes pertractatio aliqua specialis hic. Addendo specialem habere potest a pluribus apud nos DebuggerDisplay, reflective pertractatio a parte post, si in tempore hoc est in loco hoc sit dolorem sed non puto esse tanti scriptor pretium in a ordinarius tostring vocat.

Quoting / paululum effugiunt values
Vestibulum nec libris fugam cum chordarum Typographia sequitur quod potentia confundens excudere. Sed de hoc non habetis hic. Primo, nemo nostrum prior es in C # haec fugam intercluserant. Secundo, in compage etiam non evadere tostring () sic. Tertium genus est id quod reverti possit de hoc quaestio non esset hic. An hoc volo ut procer newlines # version C: \ n, vel VB version? Denique iam aliquid dictum evaluator tractandis fidibus et anonymi debugger regularis rationes spectat bonum ut ostenderentur. Datum hoc maxime de cursus est, ut appareat records bene formatae sint in debugger, ut videtur rectam aditus est ad agnosticismum et lingua, expressio evaluator scit de recordum in tuto collocaretur ipsa cum bene generibus.

conclusioni
Et nos non piget aliqua de filum aut suomet recordum membra. Quid faciet nobis melius debugger hic.

Si autem omittamus implementation tostring in abstracto records
Nos posse in motu violento Antiphona modum verae exsecutionem hoc recordum types in abstracto, quia non est qui dicitur per derived tostring implementation. In facie ejus, et non cogitare de sem non esse praecipue a medio inter quas toString, sed ut etiam possit non cogitare de sem ut non lædetur a comprehendo, et faciunt ut esset super compiler implementation simpler.

conclusioni
Custodiunt illud.

Si prior ad vocamus tostring StringBuilder.Append ad valorem types
Quod vocant de recta et non arcetur tostring in valore est, ut transire non-prima figura efficere, hoc faciam ut in artem efficere solere; et mores et est arcetur statim vocant tostring et transeamus ad filum cultro.

conclusioni
Facite.

Conantur vitare debemus vacuo spatio duplex testimonium
Exsecutionem est inanis est currently procer personam} {record. Hic statim provocat 'imperfecta' reactiones in LDT, similes trailing super commatibus.

conclusioni
Si hoc removendum est? ETIAM! ETIAM! ETIAM! ETIAM!

Magis debemus conari ut aliquet typename caput print
Hodie vocamus StringBuilder.Append (laetus (genus)) et StringBuilder.Append ( "{"), ut statim postquam separatum vocat II, quod aliud est modum fecit universis vocationem potuimus stillabunt si nos in unum. Nos sentire quod hoc satis sit, ut aliquid sit amet lingua extra compiler facere si opportunum iudicaverit. Ex compiler perspective autem est mundus potuit prius esse malum commutatione: ut esset auget magnitudinem filum mensam, quae nota est an exitus per aliqua progressio, cum tollendum unum extra modum vocatio per modum ne praecipue disposito quam primum ad primum.

conclusioni
Ad implementors. Tamen, si non credimus ita esse vel boni ideam si tanti conatus.

Reference aequalitatem breviter circuiting
Hodie nos habere modica differentia inter aequalitatem operators == /? = Implementation et ab Pares (record) modum exempli gratia, in prima quidem pristinam aequalitatem similes referat, coram delegato ab Pares ad modum. Nullum est idque efficit ut irritum sit reprehendo vivos et utique nullum non modo aperte ad operand comparabit medicos elit. Exempli gratia ex aequo agetis membrum erit, non reprehendo hac conditione. Tamen hoc modo quod isti duo habitus perficientur members stantia se diversimode habere possunt, cum operators ens in ordine ad citius vel parva magnitudine types recordum illud quod est in potentia, facere praeter plura similia. Propositum igitur est ut reprehendo referat ad aequalitatem Pares exempli quoque membrum quod primum elementum.

conclusioni
Non faciam istud. Nos referat ad aequalitatem ut youll 'quoque reprehendo operators in ut nulla sit == vitio nullitatis infecta, adde et unum ad modum exempli gratia initium. Est enim medium quod operators == reprehendo bis referat aequalitatem, sed eximius velox reprehendo et hoc sic suus 'a magnus spectat, non creditis.
</details>

**Comment by HaloFour** [comment]
202 does make sense.  This is the year that doesn't end.

**Comment by jnm2** [comment]
https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-08-24.md#should-we-omit-the-implementation-of-tostring-on-abstract-records:

> We could theoretically omit providing an implementation of this method on `abstract` record types as it will never be called by the derived `ToString` implementation.

Users can manually implement ToString, right? In which case it may be called by the derived `ToString` implementation. This seems like a good thing.

**Comment by 333fred** [comment]
> Users can manually implement ToString, right? In which case it may be called by the derived `ToString` implementation. This seems like a good thing.

That's part of the reason for keeping it, yes.
